lib/deep-mark-required/index.ts

```diff
@@ -0,0 +1,30 @@
+import { AreNonDistributiveEqual } from "../are-non-distributive-equal";
+import { HasParsablePath } from "../has-parsable-path";
+import { MarkRequired } from "../mark-required";
+import { Paths } from "../paths";
+import { Prettify } from "../prettify";
+import { UnionToTuple } from "../union-to-tuple";
+
+type RecursiveDeepMarkRequiredSinglePath<Type, KeyPath> = KeyPath extends `${infer PropertyKey}.${infer RestKeyPath}`
+  ? {
+      [Key in keyof Type]: AreNonDistributiveEqual<Key, PropertyKey> extends true
+        ? Prettify<RecursiveDeepMarkRequiredSinglePath<NonNullable<Type[Key]>, RestKeyPath>>
+        : Type[Key];
+    }
+  : KeyPath extends keyof Type
+  ? Prettify<MarkRequired<Type, KeyPath>>
+  : Type;
```

Comment:
The current implementation fails at these cases:

#### 1. Fails to preserve `null | undefined`

```ts
type Test = RecursiveDeepMarkRequiredSinglePath<{ a: { b?: 1 } | null }, "a.b">;
//   ^? type Test = { a: { b: 1 } }
```
`Test` should have ideally been `{ a: { b: 1 } | null }` but instead it's `{ a: { b: 1 } }`. And this happens because of the usage of `NonNullable` in L11.

#### 2. Doesn't work properly with unions
```ts
type Test = RecursiveDeepMarkRequiredSinglePath<{ a: { b?: 1 } | { c: 1 } }, "a.b">;
//   ^? type Test = { a: { b?: 1 } | { c: 1 } }
```
Here, we get back the same type we passed, without "b" being marked as required. And this happens because the conditional `KeyPath extends keyof Type` fails as there are no overlapping keys between `{ b?: 1 }` and `{ c: 1 }`.

<hr />

Following is a possible fix for the above mentioned issues:
```suggestion
type RecursiveDeepMarkRequiredSinglePath<Type, KeyPath> = KeyPath extends `${infer PropertyKey}.${infer RestKeyPath}`
  ? {
      [Key in keyof Type]: AreNonDistributiveEqual<Key, PropertyKey> extends true
        ? Prettify<RecursiveDeepMarkRequiredSinglePath<Type[Key], RestKeyPath>>
        : Type[Key];
    }
  : Type extends object
  ? Prettify<MarkRequired<Type, KeyPath & keyof Type>>
  : Type;
```

We can even move the conditional `Type extends object` to the top, this reads better IMO.
```ts
type RecursiveDeepMarkRequiredSinglePath<Type, KeyPath> = Type extends object
  ? KeyPath extends `${infer PropertyKey}.${infer RestKeyPath}`
    ? {
        [Key in keyof Type]: AreNonDistributiveEqual<Key, PropertyKey> extends true
          ? Prettify<RecursiveDeepMarkRequiredSinglePath<Type[Key], RestKeyPath>>
          : Type[Key];
      }
    : Prettify<MarkRequired<Type, KeyPath & keyof Type>>
  : Type;
```

Reply:
@som-sm Thank you for your detailed observations!

> 1.  Fails to preserve null | undefined

I could also reproduced them after adding unit tests in https://github.com/ts-essentials/ts-essentials/pull/408/commits/3b66c318058bc111926013a3287a56f36031713c

> We can even move the conditional Type extends object to the top, this reads better IMO.

I also liked the last suggested solution, addressed it in https://github.com/ts-essentials/ts-essentials/pull/408/commits/ca6055ce581992beb5d07a94011426aeec437ee7

Reply:
Also added unit tests with unions - https://github.com/ts-essentials/ts-essentials/pull/408/commits/e0922687ed7541ba00be72263e384d0a72164d4f

Reply:
@Beraliv LGTM!